modules/gapi/src/backends/streaming/gstreamingbackend.hpp

```diff
@@ -0,0 +1,95 @@
+//
+// Copyright 2020 Intel Corporation.
+//
+// This software and the related documents are Intel copyrighted materials,
+// and your use of them is governed by the express license under which they
+// were provided to you ("License"). Unless the License provides otherwise,
+// you may not use, modify, copy, publish, distribute, disclose or transmit
+// this software or the related documents without Intel's prior written
+// permission.
+//
+// This software and the related documents are provided as is, with no
+// express or implied warranties, other than those that are expressly
+// stated in the License.
+//
+
+#ifndef OPENCV_GAPI_COPY_KERNEL_HPP
+#define OPENCV_GAPI_COPY_KERNEL_HPP
+
+#include <opencv2/gapi/gkernel.hpp>
+#include <opencv2/gapi/streaming.hpp>
+#include "gstreamingkernel.hpp"
+
+namespace cv {
+namespace gimpl {
+
+class RMatMediaBGRAdapter : public cv::RMat::Adapter
+{
+    cv::MediaFrame m_frame;
+public:
+    RMatMediaBGRAdapter(cv::MediaFrame frame) : m_frame(frame) { };
+
+    virtual cv::RMat::View access(cv::RMat::Access a) override
+    {
+        auto view = m_frame.access(a == cv::RMat::Access::W ? cv::MediaFrame::Access::W
+                                                            : cv::MediaFrame::Access::R);
+
+        return cv::RMat::View(desc(),
+                              reinterpret_cast<uchar*>(view.ptr[0]),
+                              view.stride[0],
+                              [=](){});
+    }
+
+    virtual cv::GMatDesc desc() const override
+    {
+        const auto& desc = m_frame.desc();
+        GAPI_Assert(desc.fmt == cv::MediaFormat::BGR);
+        return cv::GMatDesc{CV_8U, 3, desc.size};
+    }
+};
+
+struct Copy: public cv::detail::KernelTag
```

Comment:
How code duplication can be avoided in this case ? Macros ?

Reply:
No macros please.

You don't have much duplication here to care about.

Reply:
So, we good?

Reply:
Yes.